src/core/QCheck2.ml

```diff
@@ -0,0 +1,1868 @@
+    if b < a then invalid_arg "Gen.int_range: high < low";
+    let origin = Option.fold ~none:a ~some:(fun origin ->
+        if origin < a
+        then invalid_arg "Gen.int_range: origin < low"
+        else if origin > b then invalid_arg "Gen.int_range: origin > high"
+        else origin) origin in
+    fun st ->
+      let Tree.Tree(n, _shrinks) = if a >= 0 || b < 0 then (
+          (* range smaller than max_int *)
+          assert (b-a >= 0);
+          Tree.map (fun n -> a + n) (int_bound (b - a) st)
+        ) else (
+          (* range potentially bigger than max_int: we split on 0 and
+             choose the itv wrt to their size ratio *)
+          let f_a = float_of_int a in
+          let ratio = (-.f_a) /. (1. +. float_of_int b -. f_a) in
+          if Random.State.float st 1. <= ratio
+          then Tree.map (fun n -> - n - 1) (int_bound (- (a+1)) st)
+          else int_bound b st
+        ) in
+      Tree.make_primitive (Shrink.int_towards origin) n
+
+  let (--) low high = int_range ~origin:low low high
+
+  let oneof (l : 'a t list) : 'a t =
+    int_range 0 (List.length l - 1) >>= List.nth l
+
+  let oneofl (l : 'a list) : 'a t =
+    int_range 0 (List.length l - 1) >|= List.nth l
+
+  let oneofa (a : 'a array) : 'a t =
+    int_range 0 (Array.length a - 1) >|= Array.get a
+
+  (* NOTE: we keep this alias to not break code that uses [small_int]
+     for sizes of strings, arrays, etc. *)
+  let small_int = small_nat
+
+  let small_signed_int : int t = fun st ->
+    if Random.State.bool st
+    then small_nat st
+    else (small_nat >|= Int.neg) st
+
+  (** Shrink towards the first element of the list *)
+  let frequency (l : (int * 'a t) list) : 'a t =
+    if l = [] then failwith "QCheck2.frequency called with an empty list";
+    let sums = sum_int (List.map fst l) in
+    if sums < 1 then failwith "QCheck2.frequency called with weight sum < 1";
+    int_bound (sums - 1)
+    >>= fun i ->
+    let rec aux acc = function
+      | ((x,g)::xs) -> if i < acc+x then g else aux (acc+x) xs
+      | _ -> failwith "QCheck2.frequency crashed for an unknown reason"
+    in
+    aux 0 l
+
+  let frequencyl (l : (int * 'a) list) : 'a t =
+    List.map (fun (weight, value) -> (weight, pure value)) l
+    |> frequency
+
+  let frequencya a = frequencyl (Array.to_list a)
+
+  let char_range ?(origin : char option) (a : char) (b : char) : char t =
+    (int_range ~origin:(Char.code (Option.value ~default:a origin)) (Char.code a) (Char.code b)) >|= Char.chr
+
+  let random_binary_string (length : int) (st : Random.State.t) : string =
+    (* 0b011101... *)
+    let s = Bytes.create (length + 2) in
+    Bytes.set s 0 '0';
+    Bytes.set s 1 'b';
+    for i = 0 to length - 1 do
+      Bytes.set s (i+2) (if RS.bool st then '0' else '1')
+    done;
+    Bytes.unsafe_to_string s
+
+  let int32 : int32 t = fun st ->
+    let x = random_binary_string 32 st |> Int32.of_string in
+    Tree.make_primitive (int32_towards 0l) x
```

Comment:
Yeah I don't mind changing it, I copied the current QCheck code that generates int32/64 there :sweat_smile: 

Note, this also looks a bit like https://github.com/c-cube/qcheck/issues/105#issuecomment-828508598 but as a first step, I can (in this PR) clean the 32/64 generators

Reply:
On second thought, I prefer **not** to do it in this PR.

The current implementation of `ui32` and `ui64` [is already like that](https://github.com/c-cube/qcheck/blob/master/src/core/QCheck.ml#L202) and nobody particularly complained

As this problem is already tracked in https://github.com/c-cube/qcheck/issues/90 and this would be transparent for users, I'd rather we keep this change for another, simpler PR later :grin: